### ISSUE TITLE
Mixing [RegularFile|Directory]Property API's with deferred task configuration forces task configuration

### ISSUE THREAD

**Issue by JLLeitschuh** [gradle/gradle#6363]
**TL;DR**: I think `Provider`needs a `flatMap` mechanism

This issue originates from this discussion with @oehme here #6336 and is also tangentially related to this issue #6361

Currently, there is no way to mix the new `TaskProvider` API introduced in `4.8` with the property configuration of Tasks introduced since `4.0`.

These are the example tasks that I'm working with for subsequent examples:
```kotlin
@Suppress("LeakingThis")
open class OutputTest : DefaultTask() {
    @get:OutputFile
    val outputFile: RegularFileProperty = newOutputFile()
}

@Suppress("LeakingThis")
open class InputTest : DefaultTask() {
    @get:InputFile
    @get:PathSensitive(PathSensitivity.RELATIVE)
    val inputFile: RegularFileProperty = newInputFile()
}
```

### Expected Behavior
There needs to be a way to `flatMap` from a `TaskProvider` to a `[RegularFile|Directory]Property` on that task.

I need a mechanism to do the following:
```kotlin
val out1 =
    tasks.register<OutputTest>("output1")

val dependencyWorks1 =
    tasks.register<InputTest>("dependencyWorks1") {
        inputFile.set(out1.flatMap { it.outputFile })
    }
```

### Current Behavior

I can't set the `inputFile` field on `InputTest` without forcing the configuration of `OutputTest`.

```kotlin
val out1 =
    tasks.register<OutputTest>("output1")

val dependencyWorks1 =
    tasks.register<InputTest>("dependencyWorks1") {
        // Notice the call to `get` which collapses the configuration of the `OutputTest` task above.
        inputFile.set(out1.map { it.outputFile }.get())
    }
```

This example compiles but is completely wrong because `out1` never gets configured:
```kotlin
val out1 =
    tasks.register<OutputTest>("output1")
val dependencyDeferred =
    tasks.register<InputTest>("dependencyDeferred") {
        out1.configure {
            inputFile.set(outputFile)
        }
    }
/*
./gradlew --dry-run dependencyDeferred
:dependencyDeferred SKIPPED

BUILD SUCCESSFUL in 0s
*/
```

### Context
The current lack of `flatMap` makes it impossible to keep task configuration lazy while also utilizing the `[RegularFile|Directory]Property` features.

### Your Environment
```
------------------------------------------------------------
Gradle 4.9
------------------------------------------------------------

Build time:   2018-07-16 08:14:03 UTC
Revision:     efcf8c1cf533b03c70f394f270f46a174c738efc

Kotlin DSL:   0.18.4
Kotlin:       1.2.41
Groovy:       2.4.12
Ant:          Apache Ant(TM) version 1.9.11 compiled on March 23 2018
JVM:          1.8.0_92 (Oracle Corporation 25.92-b14)
OS:           Mac OS X 10.13.5 x86_64
```

**Comment by mkobit** [comment]
I also requested that `flatMap` be added in https://github.com/gradle/gradle/issues/4875 , but I think both are mostly useless until https://github.com/gradle/gradle/issues/5706 is fixed.

**Comment by JLLeitschuh** [comment]
Ah, thanks, @mkobit, didn't realize I was opening a duplicate issue.

This does expose some more examples that I've been running into in my own plugin.

**Comment by adammurdoch** [comment]
The plan is that you'd use this pattern to wire tasks together:

```
val out1 =
    tasks.register<OutputTest>("output1")

val dependencyWorks1 =
    tasks.register<InputTest>("dependencyWorks1") {
        inputFile.set(out1.get().outputFile)
    }
```

In other words, if you need access to an output of a task, reference the properties of the producing task directly. Are you not able to use this pattern?

**Comment by JLLeitschuh** [comment]
Looking at what you are saying, I think I understand. It just seems very unintuitive when thinking about this from a purely functional standpoint.

If I understand correctly, the reason you are okay with calling `out.get()` is because if `InputTest` is being configured, then it must mean that `OutputTest` also needs to be configured at this point as well.

Maybe this makes sense, maybe not. 

Personally, I was under the impression we were trying to avoid calls to `TaskProvider::get` because it would force configuration.

My assumption that you should avoid the `get` method is reinforced by @oehme's comment here:

> You are calling get() on the Provider which you shouldn't. Just pass the provider into the other Property directly. This may also be the reason for the other problem above, as that example also has a get() that shouldn't be there.

https://github.com/gradle/gradle/issues/6336#issuecomment-412317709

Maybe the API semantics still need to be hashed out? From a functional standpoint adding a `FlatMap` API should function exactly the same as the sample you've provided above, unless I'm missing something.

**Comment by oehme** [comment]
@JLLeitschuh My comment was in a different context : publishing. Publications aren't lazy (yet), so calling `get()` there would always realize the taks.

**Comment by JLLeitschuh** [comment]
Ah, this makes sense. Thank you for clarifying @oehme.
